src/components/time-picker/time-picker.tsx

```diff
@@ -0,0 +1,382 @@
+                    onBlur={this.onInputBlur}
+                    onKeyDown={this.onInputKeyDown}
+                    onMouseDown={this.onInputMouseDown}
+                    onClick={this.onInputClick}
+                />
+                {!isTouch && !disabled && inputValue &&
+                    <Stepper
+                        className="stepper"
+                        disableUp={
+                            currentIsTime &&
+                            !isValidValue(currentSegmentValue + 1, currentSegment as any, ampm)
+                        }
+                        disableDown={
+                            currentIsTime &&
+                            !isValidValue(currentSegmentValue - 1, currentSegment as any, ampm)
+                        }
+                        onUp={this.createStepperHandler(+1)}
+                        onDown={this.createStepperHandler(-1)}
+                    />
+                }
+                {isTouch &&
+                    <label className="label">
+                        <input
+                            className="native-input"
+                            type="time"
+                            name={name}
+                            aria-label={label}
+                            value={this.getValue()}
+                            disabled={disabled}
+                            onFocus={this.onNantiveInputFocus}
+                            onBlur={this.onInputBlur}
+                            onChange={this.onNativeInputChange}
+                        />
+                    </label>
+                }
+            </div>
+        );
+    }
+
+    private getInitialState(props: Props) {
+        const segments = propsValueToSegments(props.value, props.format);
+        return {
+            ...segments,
+            inputValue: segmentsToInputValue(segments)
+        };
+    }
+
+    private getValue() {
+        const {hh, mm, ampm} = this.state;
+        return [
+            to24(hh || 0, ampm),
+            mm || 0
+        ].map(String).map(pad2).join(':');
+    }
+
+    private select(segment: Segment) {
+        const indexes = selectionIndexes[segment];
+        const input = this.input!;
+        const cursorPosition = input.selectionStart;
+        input.selectionStart = indexes[0];
+        input.selectionEnd = indexes[1];
+        this.setState({currentSegment: segment});
+        return input.selectionStart !== input.selectionEnd;
+    }
+    private moveSelection(step: number) {
+        const segments = Object.keys(selectionIndexes) as Segment[];
+        const index = segments.indexOf(this.state.currentSegment!);
+        const nextSegment = segments[index + step];
+        if (nextSegment) {
+            return this.select(nextSegment);
+        }
+        return false;
+    }
+
+    private updateSegmentValue(name: Segment, value: number | Ampm) {
+        const {hh, mm, ampm} = this.state;
+        this.setState({
+            [name as any]: value,
```

Comment:
no, typescript does not allow to set object with any `string` key to state.
it will throw
```js
Argument of type '{ [x: string]: string | number; inputValue: string; }' is not assignable to parameter of type 'Pick<State, "ampm" | "hh" | "mm" | "inputValue" | "focus" | "currentSegment">'
```